tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_docker.py

```diff
@@ -0,0 +1,146 @@
+        self.role = config.get("role", "agent")
+        self.cleanup_cmd = config.get("cleanup_cmd", None)
+        self.conn_cmd = config.get("conn_cmd", None)
+        self.mac = config.get("mac", self._get_rand_mac(oct_amount=4))
+        self.use_tcpdump = config.get("use_tcpdump", False)
+        self.docker_network = config.get("docker_network",
+                                         "prplMesh-net-{}".format(self.unique_id))
+
+        if self.cleanup_cmd:
+            self._run_shell_cmd(self.cleanup_cmd)
+
+        docker_run_cmd = os.path.join(rootdir, "tools", "docker", "run.sh") \
+            + " --verbose --detach --force --name {}".format(self.name) \
+            + " --mac {}".format(self.mac) \
+            + " --network {}".format(self.docker_network)
+
+        if self.role == "both":
+            # Spawn dockerized controller + agent in single device
+            self._run_shell_cmd(docker_run_cmd + " start-controller-agent")
+            time.sleep(config.get("delay", 5))
+            self.al_entity = ALEntityDocker(self.name, is_controller=True)
+        elif self.role == "controller":
+            # Spawn dockerized controller
+            self._run_shell_cmd(docker_run_cmd + " start-controller")
+            time.sleep(config.get("delay", 5))
+            self.al_entity = ALEntityDocker(self.name, is_controller=True)
+        else:
+            # Spawn dockerized agent
+            self._run_shell_cmd(docker_run_cmd + " start-agent")
+            time.sleep(config.get("delay", 5))
+            self.al_entity = ALEntityDocker(self.name, is_controller=False)
+
+        self.gw_iface = _get_bridge_interface(self.docker_network)
+
+        self.check_status()
+
+    def __del__(self):
+        if self.cleanup_cmd:
+            self._run_shell_cmd(self.cleanup_cmd)
+
+    def _get_rand_mac(self, oct_amount: int):
+        """
+        Produces MAC address with randomly generated octets
+        """
+        oct_amount = 6 if oct_amount > 6 else oct_amount
+        oct_amount = 1 if oct_amount < 1 else oct_amount
+        """
+        For some reason, sometimes randomly generated MAC causes ALEntity to throw exception.
+        The reason of that behavior wasn't clarified so far, so a temporary empiric solution
+        is provided.
+        """
+        # mac = "{:02x}".format(random.randint(0, 255))
+        mac = "00"
+        for i in range(1, oct_amount):
+            mac += ":{:02x}".format(random.randint(0, 255))
+        return mac
+
+    def _run_shell_cmd(self, cmd: str = ""):
+        """
+        Runs specified command in host's shell and saves its output into device's log
+        """
+        res = subprocess.run(("/bin/bash", "-c", cmd), stdout=subprocess.PIPE,
+                             stderr=subprocess.STDOUT, encoding="utf-8")
+        self.log_calls += cmd
+        self.log += res.stdout
+        logger.debug(res.stdout)
```

Comment:
Avoid using shell for subprocesses, instead use proper individual command-line options as a list. We do need a shell for commands specified in the config file, but there shouldn't be a need for those in our specific PrplMeshDocker class.

I think this wrapper for subprocess is not terribly useful because it's not called very often and it just adds the logging bit.

Also, instead of subprocess, I think it's better to use bft_pexpect_helper because it handles timeouts.

Reply:
> Avoid using shell for subprocesses, instead use proper individual command-line options as a list. We do need a shell for commands specified in the config file, but there shouldn't be a need for those in our specific PrplMeshDocker class.
> 
> Also, instead of subprocess, I think it's better to use bft_pexpect_helper because it handles timeouts.

Applied in #1108.

I've already tried to use _bft_pexpect_helper_, but it's functionality seems to be in development and not works as expected. So instead I replaced _subprocess_ calls by _pexpect_ ones.

> I think this wrapper for subprocess is not terribly useful because it's not called very often and it just adds the logging bit.

IMO, without that wrapper the code will be inflated by similar snippets, so I would prefer to leave it to respect "Don't repeat yourself" principle.